graphql-ballerina/duplicate_field_remover.bal

```diff
@@ -0,0 +1,181 @@
+        self.removeDuplicateSelections(selections);
+        foreach parser:Selection selection in selections {
+            if (selection.isFragment) {
+                parser:FragmentNode subFragmentNode = self.documentNode.getFragments().get(selection.name);
+                self.visitSelection(subFragmentNode.getSelections());
+            } else {
+                parser:FieldNode subFieldNode = <parser:FieldNode>selection?.node;
+                self.visitSelection(subFieldNode.getSelections());
+            }
+        }
+    }
+
+    public isolated function visitField(parser:FieldNode[] fields) {
+        self.removeDuplicateFields(fields);
+        foreach parser:FieldNode fieldNode in fields {
+            self.visitField(fieldNode.getFields());
+            self.visitFragment(fieldNode.getFragments());
+        }
+    }
+
+    public isolated function visitFragment(string[] fragments) {
+        self.removeDuplicateFragments(fragments);
+        foreach string fragment in fragments {
+            parser:FragmentNode fragmentNode = self.documentNode.getFragments().get(fragment);
+            self.visitFragment(fragmentNode.getFragments());
+            self.visitField(fragmentNode.getFields());
+        }
+    }
+
+    private isolated function removeDuplicateSelections(parser:Selection[] selections) {
+        map<parser:Selection> visitedSelections = {};
+        map<parser:Selection> visitedFragmentOnTypes = {};
+        int[] duplicateIndexes = [];
+        int i = 0;
+        while (i < selections.length()) {
+            parser:Selection subSelection = selections[i];
+            if(subSelection.isFragment) {
+                parser:FragmentNode subFragmentNode = self.documentNode.getFragments().get(subSelection.name);
+                if (visitedFragmentOnTypes.hasKey(subFragmentNode.getOnType())) {
+                    if (subFragmentNode.isInlineFragment()) {
+                        duplicateIndexes.push(i);
+                    } else {
+                        self.appendDuplicateSelections(selections[i], visitedFragmentOnTypes.get(subFragmentNode.getOnType()));
+                    }
+                } else {
+                    visitedFragmentOnTypes[subFragmentNode.getOnType()] = selections[i];
+                }
+            } else {
+                if (visitedSelections.hasKey(subSelection.name)) {
+                    self.appendDuplicateSelections(selections[i], visitedSelections.get(subSelection.name));
+                    duplicateIndexes.push(i);
+                } else {
+                    visitedSelections[subSelection.name] = selections[i];
+                }
+            }
+            i += 1;
+        }
+        foreach int index in duplicateIndexes.reverse() {
+            var remove = selections.remove(index);
+        }
+    }
+
+    private isolated function removeDuplicateFields(parser:FieldNode[] fields) {
+        string[] visitedFields = [];
+        int[] duplicateIndexes = [];
```

Comment:
We can make this a `parser:FieldNode[]` instead of an int array. Don't need to keep the indexes, instead keep the original object reference.